pkg/execution/state/redis_state/redis_state.go

```diff
@@ -243,14 +243,26 @@ func (m shardedMgr) New(ctx context.Context, input state.Input) (state.State, er
 	// atomicity to improve idempotency.
 	//
 	// In future/other metadata stores this is (or will be) transactional.
-	res := client.Do(ctx, func(c rueidis.Client) rueidis.Completed {
-		return c.B().Set().Key(
-			fnRunState.kg.Idempotency(ctx, isSharded, input.Identifier),
-		).Value("").Nx().Ex(consts.FunctionIdempotencyPeriod).Build()
-	})
-	set, err := res.AsBool()
-	if (err == nil || rueidis.IsRedisNil(err)) && !set {
-		return nil, state.ErrIdentifierExists
+	//
+	{
+		key := fnRunState.kg.Idempotency(ctx, isSharded, input.Identifier)
+		runID, err := m.idempotencyCheck(ctx, client, key, input.Identifier)
+		switch err {
+		case nil: // no-op
+		// NOTE:
+		// This will happen as part of the transition of storing empty strings for idempotency
+		// key to ULID values.
+		// So if this error is returned, we should just continue with creating a new state, since
+		// it could mean that the state is not actually created.
+		case state.ErrInvalidIdentifier: // no-op
+		default:
+			return nil, err
+		}
+
+		// If a state already exists with the idempotency key, override the input's runID and continue
+		if runID != nil && input.Identifier.RunID != *runID {
```

Comment:
If state already exists, there's a chance that a step may have already ran. Shall we return state from the state store via a lookup here?

Reply:
I've already tried that, but it causes a race condition which the idempotency testharness caught.